docs/proposals/api-group.md

```diff
@@ -0,0 +1,104 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/api-group.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API to modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+6. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified, the function should search groups in a predetermined order.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from app.APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
+
+2. handling common API objects like api.Status:
+
+  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
```

Comment:
What about having (group, version) tuples able to express a "uses" dependency relationship on another (group, version)?  That would make common cases unambiguous.  For instance, (openshift, v1beta1) uses (kube, v1beta3)?  

That would allow us to serialize in a predictable manner.  If we allowed multiple versions of the same group, that would allow clients to choose to receive their content back with heterogeneous versions.